docs: reframe Python API index page as Welcome

ecormany opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

This is the first step towards a refresh of the Python API docs. It takes the existing index page and reframes it as a Welcome page — something that many other API docs sites do.

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

## Template changes
- Hides "welcome" banner on index page.
- Hides banner explaining that this is the v2 documentation on all other pages (this should be self-evident from the sidebar nav)
## index.rst
- Renames index page "Welcome"
- New prose introduction describing beginner and advanced uses
- Reworked Getting Started section
- New explanation for sample protocol code
- Removes How It's Organized section to `writing.rst`
- Adds new links to other Opentrons resources
## writing.rst
- Imports How Protocols are Organized section from `index.rst`
- Turns Robot's Jupyter Notebook into a subsection
- Typo fixes


# Review requests

<!--
Describe any requests for your reviewers here.
-->

These changes are open for comments from anyone who contributes to or uses these docs. Please let me know what you would like to see before this PR is merged and what can be added to the to-do list for future evolutions of the Welcome page!


# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

Very low. Some reference anchors were moved, but none were renamed or removed.

Commit:
Fix social links in footer

Commit:
Merge branch 'edge' into ecormany-docs-welcome

ecormany committed:
Merge remote-tracking branch 'origin' into ecormany-docs-welcome

ecormany committed:
reframe index as Welcome page

ecormany committed:
removed stray comment in index.rst

SyntaxColoring commented on api/docs/templates/v2/layout.html:
```diff
@@ -1049,6 +1049,7 @@
     <div class="bodywrapper">
       {%- endif %}
       <div class="body" role="main">
+<!-- 
```
For cleanliness, I think we should outright delete the banner HTML, not comment it out. If we ever need to restore it, we can always fall back on the Git history.

ecormany replied:
yeah, was treading too lightly here.

SyntaxColoring commented on api/docs/v2/index.rst:
```diff
@@ -1,58 +1,38 @@
-We’ve designed it in a way we hope is accessible to anyone with basic Python and wetlab skills. As a bench scientist, you should be able to code your automated protocols in a way that reads like a lab notebook.
+As a bench scientist, you should be able to code your protocols in a way that reads like a lab notebook. You can :ref:`write a fully functional protocol <writing>` just by listing the equipment you'll use (modules, labware, and pipettes) and the exact sequence of movements the robot should make.
 
-Version 2 of the API is a new way to write Python protocols. It is more reliable, simpler, and better able to be supported. Unlike version 1, it has support for new modules like the Thermocycler. While version 1 will still recieve bug fixes, new features and improvements will land in version 2. For a guide on transitioning your protocols from version 1 to version 2 of the API, see `this article o ...
```
I agree that this version 2 note didn't belong here, and some of it is obsolete anyway, but I think parts of it are keeping around somewhere.

Should we move this to the "Versioning" page under a v2.0 heading, maybe?

* The note about version 1 still receiving bug fixes is wrong now. v1 hasn't been supported for a while. I think we should just remove that sentence.
* I think the link to the migration guide is worth keeping around. IMO that guide should actually live here on docs.opentrons.com, but that's a problem for another day. 🙂 
* The link to the article on background information and rationale is also worth keeping around, IMO.

ecormany replied:
Moved this paragraph to Versioning as suggested, deleting the bug fix notice. Making a note to work with Support on importing info from the migration guide in the future.

SyntaxColoring commented on api/docs/v2/index.rst:
```diff
@@ -1,58 +1,38 @@
-To simulate protocols on your laptop, check out :ref:`simulate-block`. When you're ready to run your script on a robot, download our latest `desktop app <https://www.opentrons.com/ot-app>`_.
+If you want to **dive right into code**, take a look at our :ref:`new-examples` page, the `Protocol Library <https://protocols.opentrons.com>`_, and the comprehensive :ref:`protocol-api-reference`.
 
+When you're ready to **try out a protocol**, you can :ref:`simulate it on your computer <simulate-block>` — regardless of whether you're connected to an OT-2 robot. To run your script on a robot, download and use our latest `desktop app <https://www.opentrons.com/ot-app>`_.
```
What do you think about the usage of the term "script" in this section? I worry it might be unfamiliar to Python newbies. Should we uniformly refer to these files as "Python protocols"?

ecormany replied:
i adopted "script" since it was used a few lines above. happy to standardize on "protocol" even if it ups repetition a bit.

SyntaxColoring commented on api/docs/v2/index.rst:
```diff
@@ -1,58 +1,38 @@
+**New to Python protocols?** Check out the :ref:`writing` page first to get a sense of the typical structure of our scripts. 
 
-To simulate protocols on your laptop, check out :ref:`simulate-block`. When you're ready to run your script on a robot, download our latest `desktop app <https://www.opentrons.com/ot-app>`_.
+If you want to **dive right into code**, take a look at our :ref:`new-examples` page, the `Protocol Library <https://protocols.opentrons.com>`_, and the comprehensive :ref:`protocol-api-reference`.
```
I'm a little bit hesitant to explicitly recommend looking at protocols.opentrons.com because many of those files do confusing hacks to work around limitations in our software. To provide configurability, they have big JSON objects concatenated into them at download time, and then they parse those JSON objects at runtime.

So people should feel free to poke around them if they're curious, but not necessarily use them as examples.

ecormany replied:
🤐 (for now)

SyntaxColoring commented on api/docs/v2/index.rst:
```diff
@@ -1,58 +1,38 @@
+The design goal of this API is to make code readable and easy to understand. A protocol, in its most basic form:
 
-If we were to rewrite this with the Python Protocol API, it would look like the following:
+1. Tells Python to load the Opentrons Protocol API.
```
If you're referring to the `from opentrons import protocol_api` statement, then technically, that's not *quite* what that line is doing. See my comment below.

SyntaxColoring commented on api/docs/v2/index.rst:
```diff
@@ -1,58 +1,38 @@
-If we were to rewrite this with the Python Protocol API, it would look like the following:
+1. Tells Python to load the Opentrons Protocol API.
+2. Provides some information about who made the protocol and what it is for.
+3. Tells the robot where to find labware, pipettes, and (optionally) modules.
```
```suggestion
3. Tells the robot where to find labware, pipettes, and (optionally) hardware modules.
```

To avoid potential confusion with Python modules.

ecormany replied:
makes sense. wasn't sure how to balance verbosity vs. ambiguity.

SyntaxColoring commented on api/docs/v2/index.rst:
```diff
@@ -1,58 +1,38 @@
+1. Tells Python to load the Opentrons Protocol API.
+2. Provides some information about who made the protocol and what it is for.
+3. Tells the robot where to find labware, pipettes, and (optionally) modules.
+4. Gives commands to the robot to manipulate its attached hardware.
```
```suggestion
4. Commands the robot to manipulate its attached hardware.
```

ecormany replied:
✅ more concise, equally clear.

SyntaxColoring commented on api/docs/v2/index.rst:
```diff
@@ -61,18 +41,17 @@ If we were to rewrite this with the Python Protocol API, it would look like the
-        'author': 'Name <email@address.com>',
-        'description': 'Simple protocol to get started using OT2',
+        'author': 'Name <opentrons@example.com>',
+        'description': 'Simple protocol to get started using OT-2',
```
```suggestion
        'description': 'Simple protocol to get started using the OT-2',
```

SyntaxColoring commented on api/docs/v2/index.rst:
```diff
@@ -84,123 +63,50 @@ If we were to rewrite this with the Python Protocol API, it would look like the
-++++++++++++++++++
-
-When writing protocols using the Python Protocol API, there are generally five sections:
+1. Loads the Opentrons Protocol API.
```
If this is referring to the `from opentrons import protocol_api`, statement, then that doesn't actually load the API. All it does it let you write your `run()` function like this:

```python
def run(protocol: protocol_api.ProtocolContext):
```

Instead of like this:
```python
def run(protocol):
```

Which lets your editor do static type-checking and autocomplete the methods on the `protocol` object. In theory, anyway. In practice, I don't know if any protocol authors actually take advantage of this.

Anyway, the protocol author could remove that `import` statement and use the simpler second form, if they wanted to. The API is always "loaded" implicitly, in other words.

ecormany replied:
Interesting, and thanks for explaining the nuanced difference! How should we explain this at a high level to a new API user? `from opentrons import protocol_api` is line 1 and _feels_ important, but the solution might be to omit explaining it here.

mcous replied:
Yeah I think maybe the best move here is to have a separate section or something that notes how to write protocols with type-checking, which is entirely optional. So I agree that we should remove it in this very first protocol example.

Maybe a "Using Type Annotations" section in `writing.rst`?

ecormany replied:
sounds like a good idea (for a future update)

SyntaxColoring commented on api/docs/v2/writing.rst:
```diff
@@ -51,6 +51,110 @@ __ https://code.visualstudio.com/
+
+When writing protocols using the Python Protocol API, there are generally five sections:
+
+1) Metadata and Version Selection
+2) Run function
+3) Labware
+4) Pipettes
+5) Commands
```
Especially with the abstract discussion of `metadata` below, should we pair this outline with a complete example?

Actually, should we just move the example that you currently have on the "Welcome" page to here? You're already doing a line-by-line breakdown there. Moving that stuff here would have the advantage of deduplicating it.

ecormany replied:
it's entirely possible that the Welcome page is over-explaining. on the other hand, i didn't want it to appear too empty. interested to hear what others think too. 

basically: is a very short Welcome better, or does it need an example?

SyntaxColoring replied:
I'm personally happy with a welcome page that's brief, contains no example, and leads readers forwards to pages where there are examples.

Another thing to consider is the limitations of our navigation. Currently, this index/intro/welcome page is kind of hard to find in the left sidebar. It's this link:

<img width="261" alt="Screen Shot 2022-04-08 at 5 40 47 PM" src="https://user-images.githubusercontent.com/3236864/162535851-75eaf869-c83a-4b0b-9fb1-51b9497f5e1c.png">

Which is easy to miss because it doesn't really look like a link, and because its text doesn't match the page's actual title. 

There might be a way to fix this. Not sure. But assuming we have to live with it, shunting details out of the welcome page means we can put them in places that will be more easily discoverable through the sidebar.

ecormany replied:
i wasn't able to get the full Welcome hierarchy into the sidebar (toctrees are weird and index pages are special) but i did manage to get the top-level link there by including `self` and moving the directive from the bottom of the source file.

SyntaxColoring reviewed (commented):
🎉 

This is looking great. Various comments/suggestions below.

ecormany committed:
"modules" -> "hardware modules" for clarity

Co-authored-by: Max Marrone <max@marrone.nyc>

ecormany committed:
concise language

Co-authored-by: Max Marrone <max@marrone.nyc>

ecormany committed:
add "the" in "the OT-2"

i discovered the batch accept feature too late 😆

Co-authored-by: Max Marrone <max@marrone.nyc>

mcous commented on api/docs/v2/index.rst:
```diff
@@ -61,18 +41,17 @@ If we were to rewrite this with the Python Protocol API, it would look like the
-        plate = protocol.load_labware('corning_96_wellplate_360ul_flat', '2')
-        tiprack = protocol.load_labware('opentrons_96_tiprack_300ul', '1')
+        plate = protocol.load_labware('corning_96_wellplate_360ul_flat', '1')
+        tiprack = protocol.load_labware('opentrons_96_tiprack_300ul', '2')
```
Something I'd really like to see our API documentation do more of is encourage some readability best-practices when using our API. One that comes up for me a lot is using named parameters with `load_labware`

```suggestion
        plate = protocol.load_labware('corning_96_wellplate_360ul_flat', location='1')
        tiprack = protocol.load_labware('opentrons_96_tiprack_300ul', location='2')
```

Same comment applies below to the `mount` argument of `load_instrument`

mcous commented on api/docs/v2/writing.rst:
```diff
@@ -51,6 +51,110 @@ __ https://code.visualstudio.com/
 
     For example, ``my_protocol.py``
 
+How Protocols Are Organized
```
I'm looking at the rendering in the [sandbox](http://sandbox.docs.opentrons.com/ecormany-docs-welcome/) and it looks like this all got captured in the `note` admonition. Do we need another newline or something?

<img width="718" alt="Screen Shot 2022-04-08 at 2 58 46 PM" src="https://user-images.githubusercontent.com/2963448/162506529-73319bac-3bb6-4321-90c3-03f6a74a07b6.png">

ecormany replied:
this is showing up on several pages, i even opened an issue for it https://github.com/Opentrons/opentrons/issues/9836 

i think the sitewide solution may be to bump our sphinx-doc version, it renders properly when i use sphinx-build 4.5 locally.

mcous reviewed (commented):
Looks like we've got a little rendering issue on the `writing.rst` page. Otherwise just got a few little comments

ecormany committed:
edits per review comments
changes to template, Welcome, and Versioning page to resolve most (but not all) comments on initial PR

ecormany committed:
added names to arguments

mcous reviewed (approved):
I really like the direction this is going, and I appreciate all your hard work on this! The punts and extra tickets you've filed sound good to me

ecormany committed:
added Welcome to side nav

ecormany committed:
removed explanation of import statement from example
